### PR TITLE
Feature/181713000 pick the default plot type and parameter values from the backend

### PR DESCRIPTION
**Pull request by upendrakumbham**
Ann data feature:
- Fetches the default plot type and its corresponding parameterisation values from the DB.
- SQL query finds the highest value of the `priority` column per each plot type, fetches that plot type and the corresponding parameterisation as a default.
- Endpoint `/json/cell-plots/{experimentAccession}/default/plot-method`
- Associated PR - [atlas-web-core](https://github.com/ebi-gene-expression-group/atlas-web-core/pull/111)

Note: This story has a dependency - we need to merge this PR once we merge [migrate-to-schema-v19](https://github.com/ebi-gene-expression-group/atlas-web-single-cell/pull/277). Otherwise, Jenkins build will break due to legacy code with latest schema. 

Here is the final output :  `{umap={"n_neighbors":15}, tsne={"perplexity":20}}`

### CODE REVIEW

File: app/src/main/java/uk/ac/ebi/atlas/experimentpage/cellplot/CellPlotDao.java [hunk 1]
```diff
@@ -125,4 +120,32 @@ public List<TSnePoint.Dto> fetchCellPlotWithExpression(String experimentAccessio
                                 rs.getDouble("expression_level"),
                                 rs.getString("cell_id")));
     }
+
+    private static final String SELECT_DEFAULT_PLOT_TYPE_AND_PARAMETERISATION =
+
```
Comment (lingyun1010) [inline app/src/main/java/uk/ac/ebi/atlas/experimentpage/cellplot/CellPlotDao.java]: Remove redundant line.

File: app/src/main/java/uk/ac/ebi/atlas/experimentpage/cellplot/CellPlotDao.java [hunk 2]
```diff
@@ -125,4 +120,32 @@ public List<TSnePoint.Dto> fetchCellPlotWithExpression(String experimentAccessio
                                 rs.getDouble("expression_level"),
                                 rs.getString("cell_id")));
     }
+
+    private static final String SELECT_DEFAULT_PLOT_TYPE_AND_PARAMETERISATION =
```
Comment (lingyun1010) [inline app/src/main/java/uk/ac/ebi/atlas/experimentpage/cellplot/CellPlotDao.java]: PLOT METHOD should be a better name here.

File: app/src/main/java/uk/ac/ebi/atlas/experimentpage/cellplot/CellPlotJsonSerializer.java [hunk 3]
```diff
@@ -112,4 +112,10 @@ private ImmutableList<ImmutableMap<String, Object>> modelForHighcharts(String se
                                                                            Set<TSnePoint> points) {
         return ImmutableList.of(ImmutableMap.of("name", seriesName, "data", points));
     }
+
+    public String fetchDefaultPlotTypeWithPlotOption(String experimentAccession) {
```
Comment (lingyun1010) [inline app/src/main/java/uk/ac/ebi/atlas/experimentpage/cellplot/CellPlotJsonSerializer.java]: Again here, PLOT METHOD might be a better name.

File: app/src/main/java/uk/ac/ebi/atlas/experimentpage/cellplot/CellPlotService.java [hunk 4]
```diff
@@ -90,4 +91,8 @@ public ImmutableSet<TSnePoint> expressionPlot(String experimentAccession,
                                         pointDto.name()))
                 .collect(toImmutableSet());
     }
+
+    public ImmutableMap<String, String> fetchDefaultPlotTypeWithPlotOption(String experimentAccession) {
```
Comment (lingyun1010) [inline app/src/main/java/uk/ac/ebi/atlas/experimentpage/cellplot/CellPlotService.java]: PLOT METHOD again.

File: app/src/main/java/uk/ac/ebi/atlas/experimentpage/cellplot/JsonCellPlotController.java [hunk 5]
```diff
@@ -107,4 +107,9 @@ public String expressionPlot(@PathVariable String experimentAccession,
                 parsePlotParametersFromRequestParameters(plotMethod, requestParams),
                 requestParams.getOrDefault("accessKey", ""));
     }
+
+    @GetMapping(value = "/default/plot-type")
+    public String defaultPlotTypeWithOption(@PathVariable String experimentAccession) {
```
Comment (lingyun1010) [inline app/src/main/java/uk/ac/ebi/atlas/experimentpage/cellplot/JsonCellPlotController.java]: PLOT METHOD again. Please also refactor the names in tests.

File: app/src/test/java/uk/ac/ebi/atlas/experimentpage/cellplot/CellPlotDaoIT.java [hunk 6]
```diff
@@ -168,6 +170,21 @@ void fetchExpressionPlotWithInvalidGeneIdHasNoExpression(String experimentAccess
                 .allSatisfy(dto -> assertThat(dto).hasFieldOrPropertyWithValue("expressionLevel", 0.0));
     }
 
+    @Test
+    void canFetchDefaultPlotTypeAndOption(){
+        assertThat(subject.fetchDefaultPlotTypeWithPlotOption(jdbcTestUtils.fetchExperimentAccessionByMaxPriority()))
+                .containsKeys("UMAP","t-SNE")
```
Comment (lingyun1010) [inline app/src/test/java/uk/ac/ebi/atlas/experimentpage/cellplot/CellPlotDaoIT.java]: Space after comma. Please check around.

File: app/src/test/java/uk/ac/ebi/atlas/experimentpage/cellplot/CellPlotDaoIT.java [hunk 7]
```diff
@@ -168,6 +170,21 @@ void fetchExpressionPlotWithInvalidGeneIdHasNoExpression(String experimentAccess
                 .allSatisfy(dto -> assertThat(dto).hasFieldOrPropertyWithValue("expressionLevel", 0.0));
     }
 
+    @Test
+    void canFetchDefaultPlotTypeAndOption(){
```
Comment (lingyun1010) [inline app/src/test/java/uk/ac/ebi/atlas/experimentpage/cellplot/CellPlotDaoIT.java]: A suggestion name for the test `fetchDefaultPlotMethodAndOptionWithExistingExperimentAccession`.

File: app/src/test/java/uk/ac/ebi/atlas/experimentpage/cellplot/CellPlotDaoIT.java [hunk 8]
```diff
@@ -168,6 +170,21 @@ void fetchExpressionPlotWithInvalidGeneIdHasNoExpression(String experimentAccess
...
+        assertThat(subject.fetchDefaultPlotTypeWithPlotOption(jdbcTestUtils.fetchExperimentAccessionByMaxPriority()))
+                .containsKeys("UMAP","t-SNE")
+                .doesNotContainKey("scanvi")
+                .containsValues("{\"n_neighbors\": 15}")
+                .containsValues("{\"perplexity\": 15}");
+    }
+
+    @Test
+    void canFetchEmptyResultsIfExperimentDoesNotHaveDefaultPlotType(){
```
Comment (lingyun1010) [inline app/src/test/java/uk/ac/ebi/atlas/experimentpage/cellplot/CellPlotDaoIT.java]: `fetchEmpty...` would be fine. The same for the other tests.

**Review by lingyun1010 — COMMENTED** [review]
Looks good and a few tiny comments about refactoring. Please take a look.

**Comment by upendrakumbham** [comment]
Hi @lingyun1010, I have fixed your review comments, please have a look.

File: app/src/main/java/uk/ac/ebi/atlas/experimentpage/cellplot/CellPlotDao.java [hunk 9]
```diff
@@ -122,7 +122,6 @@ public List<TSnePoint.Dto> fetchCellPlotWithExpression(String experimentAccessio
     }
 
     private static final String SELECT_DEFAULT_PLOT_TYPE_AND_PARAMETERISATION =
```
Comment (lingyun1010) [inline app/src/main/java/uk/ac/ebi/atlas/experimentpage/cellplot/CellPlotDao.java]: This postgres fetch string could be renamed too here.

File: app/src/main/java/uk/ac/ebi/atlas/experimentpage/cellplot/JsonCellPlotController.java [hunk 10]
```diff
@@ -108,8 +108,8 @@ public String expressionPlot(@PathVariable String experimentAccession,
                 requestParams.getOrDefault("accessKey", ""));
     }
 
-    @GetMapping(value = "/default/plot-type")
-    public String defaultPlotTypeWithOption(@PathVariable String experimentAccession) {
-        return cellPlotJsonSerializer.fetchDefaultPlotTypeWithPlotOption(experimentAccession);
+    @GetMapping(value = "/default/plotMethod")
```
Comment (lingyun1010) [inline app/src/main/java/uk/ac/ebi/atlas/experimentpage/cellplot/JsonCellPlotController.java]: I reckon we don't use upper case format usually here.

File: app/src/test/java/uk/ac/ebi/atlas/experimentpage/cellplot/CellPlotServiceTest.java [hunk 11]
```diff
@@ -110,20 +110,20 @@ void expressionPlot() {
...
                 .thenReturn(ImmutableMap.of("UMAP","{\"n_neighbors\": 15}"));
 
-        assertThat(subject.fetchDefaultPlotTypeWithPlotOption("E-CURD-4"))
+        assertThat(subject.fetchDefaultPlotMethodWithParameterisation("E-CURD-4"))
                 .containsKeys("UMAP")
                 .containsValues("{\"n_neighbors\": 15}");
     }
 
     @Test
     void canReturnEmptyResultIfThereIsNoDefaultPlotTypeAndOption(){
```
Comment (lingyun1010) [inline app/src/test/java/uk/ac/ebi/atlas/experimentpage/cellplot/CellPlotServiceTest.java]: `return...` would be fine

**Review by lingyun1010 — COMMENTED** [review]
Seems better but a bit more to be refactored.

**Review by lingyun1010 — DISMISSED** [review]
LGTM!

**Comment by lingyun1010** [comment]
The default plot method and parameterisation could be passed into the tsne plot component as part of the `content` in `ExperimentPageContentService`, so please consider if we still need an endpoint for that.

**Comment by lingyun1010** [comment]
Please also keep the json reponse as the format like `defaultPlotTypeAndParameterisation: {"t-SNE": {"perplexity": 50}, "UMAP": {"n_neighbors": 100}}`, the same as the props `plotTypesAndOptions`, so we could have consistent parsing method in the front-end.
FYI, currently the json reponse from the endpoint is `{"t-SNE":"{\"perplexity\": 50}","UMAP":"{\"n_neighbors\": 100}"}`.

**Review by lingyun1010 — DISMISSED** [review]
LGTM!

File: app/src/test/java/uk/ac/ebi/atlas/experimentpage/tabs/ExperimentPageContentServiceTest.java [hunk 12]
```diff
@@ -189,4 +195,21 @@ void anatomogramDoesNotExistForValidExperiment() {
...
+                .thenReturn(ImmutableMap.of());
+
+        assertThat(subject.fetchDefaultPlotMethodAndParameterisation("FooBar")).isEmpty();
+    }
+
+    @Test
+    void getEmptyDefaultPlotMethodAndParamsForTheValidExperiment(){
+        when(cellPlotServiceMock.fetchDefaultPlotMethodWithParameterisation("E-CURD-4"))
+                .thenReturn(ImmutableMap.of("UMAP",new Gson().fromJson("{\"n_neighbors\":100}",JsonObject.class),
+                                            "t-SNE",new Gson().fromJson("{\"perplexity\":50}",JsonObject.class)));
```
Comment (lingyun1010) [inline app/src/test/java/uk/ac/ebi/atlas/experimentpage/tabs/ExperimentPageContentServiceTest.java]: The key for `t-SNE` is different format from the one in controller.

File: app/src/main/java/uk/ac/ebi/atlas/experimentpage/cellplot/CellPlotService.java [hunk 13]
```diff
@@ -90,4 +93,32 @@ public ImmutableSet<TSnePoint> expressionPlot(String experimentAccession,
...
+            defaultPlotTypeAndOptions.put(umap, umapOptionObject);
+            defaultPlotTypeAndOptions.put(tsne, tsneOptionObject);
+        }
+        return defaultPlotTypeAndOptions.build();
+    }
+
+    private static JsonObject getMiddleElement(List plotOptions) {
+
+        if (plotOptions.size() % 2 == 0) { // even number
+            Object umapEvenItem = plotOptions.get((plotOptions.size() / 2 - 1));
```
Comment (lingyun1010) [inline app/src/main/java/uk/ac/ebi/atlas/experimentpage/cellplot/CellPlotService.java:124]: The variable name could be more general for both umap and tsne result.

**Review by lingyun1010 — DISMISSED** [review]
Please fix the tiny comments before merging.

File: app/src/test/java/uk/ac/ebi/atlas/experimentpage/tabs/ExperimentPageContentServiceTest.java (hunk 12 above)
Comment (upendrakumbham) [reply app/src/test/java/uk/ac/ebi/atlas/experimentpage/tabs/ExperimentPageContentServiceTest.java]: Fixed it.

File: app/src/main/java/uk/ac/ebi/atlas/experimentpage/cellplot/CellPlotService.java (hunk 13 above)
Comment (upendrakumbham) [reply app/src/main/java/uk/ac/ebi/atlas/experimentpage/cellplot/CellPlotService.java:124]: I optimised the code a bit. please have a look.

**Review by lingyun1010 — DISMISSED** [review]
LGTM!

**Review by lingyun1010 — APPROVED** [review]

**Pull request merged**